Move shared feeds to new rail and side navigation

danielevalverde opened the pull request:
## Description

- Create a new rail item for shared feed home
- Remove shared feeds from the Tipline navigation list
- Add a new icon to the sidebar that opens a new side bar
- Add counter to the section title (number of shared feeds)
- Extract FeedsComponent from ProjectsComponent
- Add new FeedPage

Reference: CV2-3927

## Type of change

- [ ] Performance improvement and/or refactoring (non-breaking change that keeps existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Security mitigation or enhancement
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Automated test (add or update automated tests)

## How has this been tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can verify the changes. Please describe whether or not you implemented automated tests.

## Things to pay attention to during code review

Please describe parts of the change that require extra attention during code review, for example:

- File FFFF, line LL: This refactoring does this and this. Is it consistent with how it’s implemented elsewhere?
- Etc.

## Checklist

- [ ] I have performed a self-review of my own code
- [ ] I've made sure my branch is runnable and given good testing steps in the PR description
- [ ] I considered secure coding practices when writing this code. Any security concerns are noted above.
- [ ] I have commented my code in hard-to-understand areas, if any
- [ ] I have made needed changes to the README
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I implemented any new components, they are self-contained, their `propTypes` are declared and they use React Hooks and, if data-fetching is required, they use Relay Modern with fragment containers
- [ ] If my components involve user interaction - specifically button, text fields, or other inputs - I have added a [BEM-like class name](https://meedan.atlassian.net/wiki/spaces/ENG/pages/1327628289/Naming+conventions+for+interactive+elements) to the element that is interacted with
- [ ] To the best of my knowledge, any new styles are applied according to the design system
- [ ] If I added a new external dependency, I included a rationale for doing so and an estimate of the change in bundle size (e.g., checked in https://bundlephobia.com/)

caiosba commented on src/app/components/drawer/DrawerRail.js:
```diff
@@ -48,14 +49,29 @@ const DrawerRail = (props) => {
+    // if drawer is open and we click on the other button, change it
+    if (drawerOpen && currentDrawerType !== newDrawerType) {
+      onDrawerTypeChange(newDrawerType);
+      onDrawerOpenChange(drawerOpen);
```
If it didn't change, is this call really needed? I'm thinking about useless renders.

danielevalverde replied:
:white_check_mark:

caiosba commented on src/app/components/drawer/DrawerRail.js:
```diff
@@ -90,6 +106,9 @@ const DrawerRail = (props) => {
             >
               <QuestionAnswerIcon />
             </Link>
+            <button type="button" className={`${styles.railIconButton} ${drawerOpen ? 'side-navigation__toggle-open' : 'side-navigation__toggle-closed'}`} id="side-navigation__toggle" onClick={() => setDrawerOpenChange('feed')}>
```
The `ID` of this element is the same as the `button` above... it's better to differentiate. If this selector is needed for any functionality, I suggest to change it to a class.

danielevalverde replied:
:white_check_mark:

caiosba commented on src/app/components/drawer/Projects/FeedsComponent.js:
```diff
@@ -0,0 +1,188 @@
+    return (inStore === 'true');
+  };
+
+  // eslint-disable-next-line
+  console.log("feedss: ", feeds)
```
Please remove these lines :-P

danielevalverde replied:
:white_check_mark:

caiosba commented on src/app/components/drawer/Projects/FeedsComponent.js:
```diff
@@ -0,0 +1,188 @@
+            <FormattedMessage
+              tagName="span"
+              id="projectsComponent.sharedFeeds"
+              defaultMessage="Colaborating [{feedsLength}]"
```
Typo: Missing an "L" in "Collaborating"

danielevalverde replied:
:white_check_mark:

caiosba commented on src/app/components/drawer/Projects/index.js:
```diff
@@ -3,28 +3,35 @@ import React from 'react';
+          feeds={feeds.map(f => ({ ...f, title: (f.name || f.feed?.name), dbid: (f.feed_id || f.dbid) }))}
+        />
+        :
+        <FeedsComponent
```
I think it's safer and more clear to explicitly check for `drawerType === 'feed'` here, so:

```jsx
if (drawerType === 'default') {
  return (<ProjectsComponent ... />);
} else if (drawerType === 'feed') {
  return (<FeedsComponent ... />);
}
return null;
```

danielevalverde replied:
:white_check_mark:

brianfleming reviewed (changes requested):
nice start @danielevalverde ! very excited for this feature. Here are a bunch of stuff to work on to start to polish this up:

- [ ] When not the active item, the new shared feed button incorrectly has a background color. If you look at this screenshot, I am on a settings page, so the shared feed menu item should look like the tipline one, but you can see it has a bg color
<img width="114" alt="Screenshot 2024-01-12 at 3 00 35 PM" src="https://github.com/meedan/check-web/assets/418001/a45d3284-7624-4ca7-b7e1-96a120cdfcf3">

- [ ] Can you add a `title=` attribute to this new item, as you can see in the screenshot, the other ones, including the Tipline have one:
<img width="118" alt="Screenshot 2024-01-12 at 3 07 50 PM" src="https://github.com/meedan/check-web/assets/418001/4592aff6-3edb-460e-8856-d1d87aaa0f4e">

- [ ] Deep linking directly to a feed shows the correct shared feed, but the new item in the rail is not active, and the drawer is showing the tipline content:


https://github.com/meedan/check-web/assets/418001/b18c1e88-a278-423a-8823-5302d9578c3d

- [ ] The new item should not toggle the collapse of the content, and you can see in this video, when you then click the collapse and expand, it shows the tipline content:

https://github.com/meedan/check-web/assets/418001/016269e1-f171-4cf6-81f3-52305b8a9d30

- [ ] When you go to a shared feed, then try to go back to the tipline, you get the All Items correctly, but the drawer content still shows the shared feeds:


https://github.com/meedan/check-web/assets/418001/614008fe-8207-45c7-806a-3abd9a32efde

- [ ] This one is my fault, something I overlooked. When you click on the new side rail item, you should be taken somewhere, right now it just shows the drawer content. The issue is, the other items behave differently. So here is what I am thinking:
  - If the user has shared feeds, then lets take them to the 1st one in the list
  - If the user does not have any shared feeds, take them to the create shared feed page
  - if those ^^ options are not as easily achievable, then we can create a main landing page for all user's shared feeds, if this is chosen, I can help with designs if you can just get the logic in place (cc @caiosba )

caiosba reviewed (changes requested):
Great stuff Dani! Pretty clean solution. I left a few comments, plus some other things:

- When we go straight to a feed route, it's important to show the feed sidebar, not the default one - I think we already have a check for that, we just need to move it to a different condition and action: https://github.com/meedan/check-web/blob/feature/CV2-3927-move-shared-feeds-to-new-rail-and-side-navigation/src/app/components/drawer/DrawerRail.js#L79. So, if `isFeedPage`, we `onDrawerTypeChange('feed')`, etc.
- Noting this weird white line on the "create feed" button:
![Captura de tela de 2024-01-12 17-08-03](https://github.com/meedan/check-web/assets/117518/46fbd08c-b8ee-4270-859e-0eee571911d0)
- This case didn't work for me:
1. Go straight to "all items" page
2. Click on the "feeds" icon: the sidebar loads but the feeds icon doesn't get blue (while the "all items" icon stays blue)
3. Click again on the "all items" icon: the sidebar doesn't change to default

caiosba commented:
Ah, just one more thing @danielevalverde - please run integration tests against this branch once you're done.

brianfleming commented:
> > This one is my fault, something I overlooked. When you click on the new side rail item, you should be taken somewhere, right now it just shows the drawer content. The issue is, the other items behave differently. So here is what I am thinking:
> > If the user has shared feeds, then lets take them to the 1st one in the list
> > If the user does not have any shared feeds, take them to the create shared feed page
> > if those ^^ options are not as easily achievable, then we can create a main landing page for all user's shared feeds, if this is chosen, I can help with designs if you can just get the logic in place (cc @caiosba )
> 
> I think the third option is the best. The second is the easiest, so, in order to keep a simpler scope in this ticket, I would advocate for that in the scope of this ticket.

Ok, sounds good @caiosba, chime in if anything sounds off with the following...

@danielevalverde when the user clicks on the new side rail, lets take them to a new landing page, maybe something like

`https://checkmedia.org/{{WORSPACE}}/feeds`

I think this would match our pattern. For now we can just put in a static landing page, maybe we will put some content about shared feeds on it...I will work on designs and content and get back to you.

caiosba commented:
Sounds good @brianfleming

danielevalverde commented on localization/react-intl/src/app/components/drawer/Projects/FeedsComponent.json:
```diff
@@ -0,0 +1,22 @@
+  {
+    "id": "projectsComponent.sharedFeeds",
+    "description": "Feeds of content shared across workspaces",
+    "defaultMessage": "Colaborating [{feedsLength}]"
```
:white_check_mark:

danielevalverde commented on src/app/components/drawer/DrawerRail.js:
```diff
@@ -48,14 +49,29 @@ const DrawerRail = (props) => {
+    // if drawer is open and we click on the same button that opened it, close it
+    } else if (drawerOpen && currentDrawerType === newDrawerType) {
+      onDrawerOpenChange(!drawerOpen);
+      onDrawerTypeChange(newDrawerType);
```
refactored to just call the `onDrawerOpenChange` when clicking on the sidebar-toggle as Brian's comment: "The new item should not toggle the collapse of the content" 
and call a new function `setDrawerTypeChange` when clicking on the "feed" icon or "tipline" icons

danielevalverde commented:
Thank you for all the detailed comments @brianfleming and @caiosba 
I think I  applied almost all the requested changes, It is missing only the design for the static landing page when clicking on the "shared feeds" icon, but please let me know if I'm still missing anything.

caiosba commented on src/app/components/drawer/DrawerRail.js:
```diff
@@ -32,42 +33,63 @@ const messages = defineMessages({
   const isSettingsPage = /\/settings\/[a-zA-Z0-9]+/.test(testPath);
   const isMediaPage = /\/media\/[0-9]+/.test(testPath);
-  const isFeedPage = /\/feed\/[0-9]+\/(request|cluster)\/[0-9]+/.test(testPath);
+  const isFeedPage = /\/team-2408\/feed\//.test(testPath);
```
Hard-coded slug - nevermind, I fixed it here.

caiosba commented on src/app/components/drawer/Projects/index.js:
```diff
@@ -3,28 +3,38 @@ import React from 'react';
+          currentUser={props.me}
+          team={props.team}
+          savedSearches={props.team.saved_searches.edges.map(ss => ss.node)}
+          feeds={feeds.map(f => ({ ...f, title: (f.name || f.feed?.name), dbid: (f.feed_id || f.dbid) }))}
```
I guess we can remove this line now.

caiosba reviewed (approved):
Thanks @danielevalverde for addressing all comments! I fixed the two minor things I commented. Let's wait for integration tests and @brianfleming's review and then we should be good to merge.

brianfleming commented:
I will get some content for the new page today. I just need to coordinate with the product team.
However, I don't think we can merge this until that happens. There are too many cases where the button does nothing right now, and if we released this, it would just be confusing. So @danielevalverde you can go ahead and put the logic in place with just a blank page for now if you like. My goal is for the new page just to be static content for now...kinda like an informative ad for shared feeds.

caiosba commented:
Thanks, @brianfleming ... let's hold on merging until we have that page implemented.

brianfleming reviewed (changes requested):
@danielevalverde ok, the copy/layout for the shared feeds landing page is ready. You can see it here:

[https://www.figma.com/file/N1W1p7anE8xxekD7EyepVE/Shared-Feeds?type=design&node-id=5471-7092&mode=design&t=11yO69sOBsb3QkUH-4](https://www.figma.com/file/N1W1p7anE8xxekD7EyepVE/Shared-Feeds?type=design&node-id=5471-7092&mode=design&t=11yO69sOBsb3QkUH-4)

Its just static text with 1 button, which should take the user to the create shared feed page.

Let me know if you have any questions, or if I can help out in any way

danielevalverde commented:
> @danielevalverde ok, the copy/layout for the shared feeds landing page is ready. You can see it here:
> 
> https://www.figma.com/file/N1W1p7anE8xxekD7EyepVE/Shared-Feeds?type=design&node-id=5471-7092&mode=design&t=11yO69sOBsb3QkUH-4
> 
> Its just static text with 1 button, which should take the user to the create shared feed page.
> 
> Let me know if you have any questions, or if I can help out in any way

Hey, @brianfleming :)
I implemented the new FeedPage, please review it and let me know if it is still missing anything

brianfleming commented on src/app/components/feed/FeedPageContent.js:
```diff
@@ -0,0 +1,131 @@
+  return (
+    <>
+      <div className={styles['feed-page-header']}>
+        <div>
```
@danielevalverde you don't really need this extra wrapping div, doesn't look like its doing anything really

brianfleming commented on src/app/components/feed/FeedPageContent.js:
```diff
@@ -0,0 +1,131 @@
+    <h2 className="typography-subtitle2">
+      {title}
+    </h2>
+    <p className="typography-body1">
```
@danielevalverde you don't need the className on `<p>` tags, they automatically get the correct css from global typography.css file

brianfleming commented on src/app/components/feed/FeedPageContent.js:
```diff
@@ -0,0 +1,131 @@
+            }
+            iconLeft={<AddIcon />}
+            variant="contained"
+            size="small"
```
@danielevalverde can you change this to `size="default"` in order to match the figma designs?

brianfleming reviewed (changes requested):
Looking great @danielevalverde ! I have added a couple SUPER small nit picky things in comments in the code, but the one real visual change is to make sure the title and icon matches the other pages in the app. I have put together a quick side-by-side comparison below. You will see:
- The shared feed title and icon is "higher" up on the page
- the shared feed icon is not aligned the same, and has a different color than "All Items"

So just a couple minor tweaks to make. Easiest way to check is if you have both pages open in separate tabs, when you go back and forth, the content should be in the same location on that page. Let me know if I can add a better explanation.


<img width="1529" alt="Screenshot 2024-01-23 at 9 43 12 AM" src="https://github.com/meedan/check-web/assets/418001/7bfa70cc-69cc-46c8-ab76-657783a47235">

danielevalverde commented:
> Looking great @danielevalverde ! I have added a couple SUPER small nit picky things in comments in the code, but the one real visual change is to make sure the title and icon matches the other pages in the app. I have put together a quick side-by-side comparison below. You will see:
> 
> * The shared feed title and icon is "higher" up on the page
> * the shared feed icon is not aligned the same, and has a different color than "All Items"
> 
> So just a couple minor tweaks to make. Easiest way to check is if you have both pages open in separate tabs, when you go back and forth, the content should be in the same location on that page. Let me know if I can add a better explanation.
> 
> <img alt="Screenshot 2024-01-23 at 9 43 12 AM" width="1529" src="https://private-user-images.githubusercontent.com/418001/298980368-7bfa70cc-69cc-46c8-ab76-657783a47235.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYwOTYwMjgsIm5iZiI6MTcwNjA5NTcyOCwicGF0aCI6Ii80MTgwMDEvMjk4OTgwMzY4LTdiZmE3MGNjLTY5Y2MtNDZjOC1hYjc2LTY1Nzc4M2E0NzIzNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDEyNFQxMTI4NDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02Njk4YjU3M2VjODM2YTdlMThiZTMxYzA0MzA0MjA0YzUwMGFiYmM2YWUwMjgyYTJjOGMxYTAxMmU0Zjk2MTE3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.wBpyLHF82eIP-IcrbPgopzuw55cbT8m5z9HgPBJlk_I">

Thanks for the details @brianfleming :) 
I implemented the requested changes, please review it and let me know if it is still missing anything

brianfleming reviewed (approved):
Thanks for all the awesome changes @danielevalverde the shared feeds look great in their new home!

![home](https://github.com/meedan/check-web/assets/418001/91fcb34f-1106-4270-aa55-409f404b633e)

caiosba commented on src/app/components/feed/FeedPage.js:
```diff
@@ -0,0 +1,35 @@
+        }}
+        render={({ error, props }) => {
+          if (!error && props) {
+            return <FeedPageContent routeParams={routeParams} permissions={props?.team.permissions} slug={props?.team.slug} />;
```
Non-blocker: No need to add `?` to `props` since the conditional above checks that `props` exists. I encourage making this change so the code seems more predictable.

caiosba commented on src/app/components/feed/FeedPageContent.js:
```diff
@@ -0,0 +1,129 @@
+          <div className={cx(styles['feed-header-title'])}>
+            <h6>
+              <FeedIcon />
+              Shared Feeds
```
This should be a `<FormattedMessage />`.

caiosba commented on src/app/components/feed/FeedPageContent.js:
```diff
@@ -0,0 +1,129 @@
+  );
+};
+
+export default FeedPageContent;
```
Please add `PropTypes` to `FeedPageContent`.
